Adds Ctdbp transforms for pathfinder

swarbhanu opened the pull request:
- Adds the CTDBP L0, L1, L2 salinity and L2 density transforms.
- Note that there is one L0 transform, one L1 transform and two L2 transforms.
- Adds integration tests: one each for L0 and L1, and another integration test which links L0, L1 and the two L2 (density and salinity) transforms together in a chain and checks that the granules published by each is the right kind.
- For the test, we are using fictitious values of calibration coefficients. If we later know about the calibration coefficients used in the real world, maybe those could be imported later for the test. Note that these CTDBP transforms are not going to stay for long. They are only for the pathfinder. Pending changes to the transform structure, they might be replaced.
- What may be added later is more assertions to check that the algorithms have been performed. So far we are only checking that the granules published by each transform contains data for the appropriate parameters.
- Also, in test_ctd_transforms.py, there were lines that were unnecessary, so they have been removed.

swarbhanu committed:
Added the python files for ctdbp

swarbhanu committed:
Wrote much of the ctdbp L0 and L1 transforms. But some fixing left for them

swarbhanu committed:
ctdbp L0 transform publishes granules each with c, t, p values

swarbhanu committed:
Wrote lot of the CTDBP transform code. Still loose ends may need to be tied

swarbhanu committed:
corrected wrong code

swarbhanu committed:
fixed some

swarbhanu committed:
Added code to test_ctdbp_L0 and could launch  the transform

swarbhanu committed:
removed an unrequired line in test_ctd_transforms

swarbhanu committed:
got the test to the point that the ctdbp L0 transform is able to receive packets and trying to process it. need to sort correct param dict to use

swarbhanu committed:
some more work

swarbhanu committed:
fixed a bug

swarbhanu committed:
fixed test_ctdbp_L0.py

swarbhanu committed:
fixed several things

swarbhanu committed:
Quite a bit of work on the test for ctdbp chain

swarbhanu committed:
refactored the test for chain of ctdbp transforms

swarbhanu committed:
simplified ctd L0 algorithm method

swarbhanu committed:
added some mods to the transforms

swarbhanu committed:
using the ctdbp_cdef_sample for the L0 transform

swarbhanu committed:
more work

swarbhanu committed:
now for the test, running own param dict. test for ctdbp_L0 working

swarbhanu committed:
fixed test for ctdbp_L1 transform. Test working

swarbhanu committed:
worked on test_ctdp_chain

swarbhanu committed:
some more fixing of chain test

swarbhanu committed:
All ctdbp tests working now

swarbhanu committed:
Added some checking that the appropriate parameters are being published within granules by the different transforms

swarbhanu committed:
corrected algorithm used and added cleanups

MauriceManning commented on ion/processes/data/transforms/ctdbp/ctdbp_L2_salinity.py:
```diff
@@ -0,0 +1,94 @@
+
+        log.debug("CTDBP L2 salinity transform publishing granule with record dict: %s", granule.record_dictionary)
+
+        granule.data_producer_id=self.id
```
is there a reason that this is only set in the L2 transform?

lukecampbell replied:
I think that originally we were going to use the data_producer_id to assist in tracking provenance but we shifted to using associations and data acquisition to track provenance. Also, when the ingestion framework was in its infancy the data_producer_id was used to determine which coverage the data belonged to before streams were readily identifiable. 

I don't believe we need to set data_producer_id anymore.

MauriceManning commented on ion/processes/data/transforms/ctdbp/ctdbp_L2_salinity.py:
```diff
@@ -0,0 +1,94 @@
+
+        sal_value = SP_from_cndr(conductivity * 10, t=temperature, p=pressure)
+
+#        sal_value = SP_from_cndr(r=conductivity/cte.C3515, t=temperature, p=pressure)
```
add a comment if this version of the calc needs to be saved so the next person working on the code knows why.

lukecampbell commented on ion/processes/data/transforms/ctdbp/ctdbp_L1.py:
```diff
@@ -0,0 +1,230 @@
+
+        #------------  Computation -------------------------------------
+        freq = (CONDWAT_L0 / 256) / 1000
+        CONDWAT_L1 = (g + h * freq**2 + I * freq**3 + j * freq**4) / (1 + CTcor * TEMPWAT_L1 + CPcor * PRESWAT_L1)
```
Can you place in-comment citations for the numerical models?  Example: https://github.com/lukecampbell/pygsw/blob/master/pygsw/practical_salinity.pyx#L191

In this case it only needs to be reference either the DPS or the document that explains why we're computing it this way.

swarbhanu replied:
Good point. I have now added the reference of the spreadsheet with the calculations (that Chris Winguard had placed there for the ctdbp transform algorithms to use). Tim A wanted the ctdbp transforms to be based on that spreadsheet.

lukecampbell commented on ion/processes/data/transforms/ctdbp/test/test_ctdbp_L0.py:
```diff
@@ -0,0 +1,250 @@
+        pc_list = []
+        for pc_k, pc in pdict.iteritems():
+            pc_list.append(self.dataset_management.create_parameter_context(pc_k, pc[1].dump()))
+
```
In tests like this please use addCleanup to mop up the test resources:

``` python
ctxt_id = self.dataset_management.create_parameter_context(pc_k, pc[1].dump())
seld.addCleanup(self.dataset_management.delete_parameter_context,ctxt_id)
```

swarbhanu committed:
Merge branch 'master' into ctdbp_transforms_for_pathfinder

swarbhanu committed:
made changes according to Maurice and Luke's comments

swarbhanu commented:
Added cleanups (Luke suggested) and removed a commented legacy computation algorithm. The new commits pushed in are according to  Maurice and Luke's comments.
